compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/hashing/Hasher.java

```diff
@@ -0,0 +1,163 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.graalvm.compiler.lir.hashing;
+
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.TreeSet;
+
+import org.graalvm.compiler.lir.gen.ArithmeticLIRGenerator;
+
+import jdk.vm.ci.meta.JavaConstant;
+import jdk.vm.ci.meta.Value;
+
+/**
+ * This class holds a hash function at a specific cardinality and min value (lowest key). The
+ * cardinality is the required size of the hash table to make the hasher injective for the provided
+ * keys.
+ */
+public final class Hasher {
+
+    /**
+     * Tries to find a hash function without conflicts for the provided keys.
+     *
+     * @param keys the keys
+     * @param minDensity the minimum density of the switch table. Used to determine the maximum
+     *            cardinality of the hash function
+     * @return an optional hasher
+     */
+    public static Optional<Hasher> forKeys(JavaConstant[] keys, double minDensity) {
+        if (keys.length <= 2) {
+            return Optional.empty();
+        } else {
+            int maxCardinality = (int) Math.round(keys.length / minDensity);
+            assertSorted(keys);
+            TreeSet<Hasher> candidates = new TreeSet<>(new Comparator<Hasher>() {
+                @Override
+                public int compare(Hasher o1, Hasher o2) {
+                    int d = o1.cardinality - o2.cardinality;
+                    if (d != 0) {
+                        return d;
+                    } else {
+                        return o1.effort() - o2.effort();
+                    }
+                }
+            });
+            long min = keys[0].asLong();
+            for (HashFunction f : HashFunction.instances()) {
+                for (int cardinality = keys.length; cardinality < maxCardinality; cardinality++) {
+                    if (isValid(keys, min, f, cardinality)) {
+                        candidates.add(new Hasher(f, cardinality, min));
+                        break;
+                    }
+                }
+            }
+            if (candidates.isEmpty()) {
+                return Optional.empty();
+            } else {
+                return Optional.of(candidates.first());
+            }
+        }
+    }
+
+    private static void assertSorted(JavaConstant[] keys) {
+        for (int i = 1; i < keys.length; i++) {
+            assert keys[i - 1].asLong() < keys[i].asLong();
+        }
+    }
+
+    private static boolean isValid(JavaConstant[] keys, long min, HashFunction function, int cardinality) {
+        Set<Integer> seen = new HashSet<>(keys.length);
+        for (JavaConstant key : keys) {
+            int hash = function.apply(key.asLong(), min) & (cardinality - 1);
+            if (seen.contains(hash)) {
```

Comment:
Can't this be simplified with the else removed?
```suggestion
            if (!seen.add(hash)) {
```